controlpanel/frontend/forms.py

```diff
@@ -49,53 +64,53 @@ def __init__(self, *args, **kwargs):
 
     def clean(self):
         cleaned_data = super().clean()
-
-        if cleaned_data['connect_bucket'] == "new" and not cleaned_data.get('new_datasource_name'):
-            self.add_error('new_datasource_name', "This field is required.")
-
-        if cleaned_data['connect_bucket'] == "existing" and not cleaned_data.get('existing_datasource_id'):
+        connect = cleaned_data['connect_bucket']
+        new_datasource = cleaned_data.get('new_datasource_name')
+        existing_datasource = cleaned_data.get('existing_datasource_id')
+
+        if connect == "new":
+            if new_datasource:
+                try:
+                    S3Bucket.objects.get(name=new_datasource)
+                    self.add_error(
+                        f"Datasource named {new_datasource} already exists"
+                    )
+                except S3Bucket.DoesNotExist:
+                    pass
+
+            else:
+                self.add_error('new_datasource_name', "This field is required.")
+
+        if connect == "existing" and not existing_datasource:
             self.add_error('existing_datasource_id', "This field is required.")
 
         return cleaned_data
 
     def clean_repo_url(self):
-        github_base_url = "https://github.com/"
         value = self.cleaned_data['repo_url']
-
-        if not value.startswith(github_base_url):
-            raise ValidationError("Invalid Github repository URL")
-
-        repo_name = value[len(github_base_url):]
-        org, _ = repo_name.split("/", 1)
-
-        if org not in settings.GITHUB_ORGS:
-            orgs = ", ".join(settings.GITHUB_ORGS)
-            raise ValidationError(
-                f"Unknown Github organization, must be one of {orgs}",
-            )
-
+        repo_name = value.replace("https://github.com/", "", 1)
         repo = get_repository(self.request.user, repo_name)
         if repo is None:
             raise ValidationError(
                 f"Github repository not found - it may be private",
             )
 
-        return value
+        try:
+            app = App.objects.get(repo_url=value)
```

Comment:
You could use something like `if App.objects.filter(repo_url=value).exists()` instead of `get()` and remove the "record doesn't exist" exception handling.

```python
if App.objects.filter(repo_url=value).exists():
    raise ValidationError(f"App already exists for this repository URL")
```

See: [Django's `exists()`](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#exists)

Reply:
True. Although it is supposedly more Pythonic to catch exceptions. In this case, I think you are correct - I'm not using the value of `app` after the exception is caught or not, so `.exists()` expresses intent more precisely.